### ISSUE TITLE
Pretraining on 256x256 and 512x512, recommended inference on 224x224

### ISSUE THREAD

**Issue by GeorgeBatch** [mahmoodlab/UNI#6]
Dear authors,

In the paper (and on HuggingFace), you mention that you used a dataset "composed of 75,832,905 [256×256] and 24,297,995 [512×512] histology images at 20× resolution". However, in the example code on GitHub and HuggingFace, you suggest using patch size 224x224 (224 is in the model name too: `assets/ckpts/vit_large_patch16_224.dinov2.uni_mass100k/`). Can you please explain why? Can the model accept other patch sizes?

Many thanks,
George

**Comment by Richarizardd** [comment]
Hi @GeorgeBatch - the `img_size` and `patch_size` is used to define the shape of `pos_embed` in the model. To clarify terminology, whereas a "patch" in slide-level tasks refers to $224 \times 224$ tissue-containing regions in a gigapixel "img", the "patch" here refers to the $16 \times 16$ tokens in a $224 \times 224$ "img". 

If you load the UNI checkpoint, it corresponds to 197 positional token embeddings. `img_size` = 224. `patch_size` = 16. Number of tokens + `[CLS]` token = $(224/16)^2 + 1 = 197$. In the DINOv2 model published by Meta, the shape of `pos_embed` is 1370, corresponding to $(518/14)^2 + 1 = 1370$, where `img_size` = 518 and `patch_size` = 14, which suggests that the model was: (1) defined with an `img_size` = 518 at start of training, (2) they trained new `pos_embed` during high-res FT. In our implementation of DINOv2 high-res FT, we followed the original recipe as closely as possible (configuration for $224^2$ pretraining), and during high-res FT (configuration not available), kept the `img_size` = 224. `patch_size` = 16 the same with `pos_embed` linearly interpolated.

**To answer your question:** Yes, you can still use the model for extracting high-resolution patch features (`dynamic_img_size` needs to be set to `True`). The `img_size` and `patch_size` is used to define the model architecture, but does not limit the image resolutions taken as input. In our evaluation, we still found better performance when extracting from $512^2$ images than $256^2$ images. I would not recommend extracting beyond $2000 \times 2000$, as you may encounter NaN errors in your feature embedding.

Regarding ablation experiments (trying pathology-specific image normalization, `SwiGLU` FFN, register tokens, `patch_size`=14, and other DINOv2 configurations suggested by Meta), due to our limited computing budget, we can not exhaustively explore all of these components (beyond the core experiments we wanted to perform in ablating model size, data size, SSL algorithms). For the past several months, we have been working on a UNIv2, so stay tuned! :^)

**Comment by GeorgeBatch** [comment]
Thank you for the explanation!

**Comment by matthew-lee1** [comment]
Hello, I know this is already closed but just curious if you tried images smaller than 224x224 as well? Let's say 112x112? Thanks!

**Comment by NazaninMn** [comment]
I have the same question. Thanks for your amazing work

**Comment by lsnls** [comment]
Hello, @Richarizardd, during the UNI training, `256×256` images were used, but the model architecture only supported `224×224` images. So I'm confused that what kind of data preprocessing has the image undergone?

Thanks and kind regards!

**Comment by nileshkumar0726** [comment]
@vadori Did you get answer to your question? Also, @Richarizardd are the 800 tissue patches picked randomly from the possible candidates?

**Comment by vadori** [comment]
@nileshkumar0726 No, I am sorry. In one of my recent works (https://arxiv.org/abs/2502.02471) I used interpolation to adjust the patch embeddings and the Conv2D projection layer. I noted at the end of the **Discussion** section that resizing is an alternative, and the performance impact of the two approaches remains an open question.
